features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/TimeseriesFetchStrategy.java

```diff
@@ -0,0 +1,533 @@
+import java.util.stream.Collectors;
+
+import org.opennms.core.sysprops.SystemProperties;
+import org.opennms.integration.api.v1.timeseries.Aggregation;
+import org.opennms.integration.api.v1.timeseries.Sample;
+import org.opennms.integration.api.v1.timeseries.TimeSeriesFetchRequest;
+import org.opennms.integration.api.v1.timeseries.immutables.ImmutableMetric;
+import org.opennms.integration.api.v1.timeseries.immutables.ImmutableTimeSeriesFetchRequest;
+import org.opennms.netmgt.dao.api.ResourceDao;
+import org.opennms.netmgt.measurements.api.FetchResults;
+import org.opennms.netmgt.measurements.api.MeasurementFetchStrategy;
+import org.opennms.netmgt.measurements.model.QueryMetadata;
+import org.opennms.netmgt.measurements.model.QueryNode;
+import org.opennms.netmgt.measurements.model.QueryResource;
+import org.opennms.netmgt.measurements.model.Source;
+import org.opennms.netmgt.measurements.utils.Utils;
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.OnmsResource;
+import org.opennms.netmgt.model.ResourceId;
+import org.opennms.netmgt.model.ResourceTypeUtils;
+import org.opennms.netmgt.model.RrdGraphAttribute;
+import org.opennms.netmgt.timeseries.impl.TimeseriesStorageManager;
+import org.opennms.netmgt.timeseries.integration.aggregation.NewtsLikeSampleAggregator;
+import org.opennms.newts.api.Measurement;
+import org.opennms.newts.api.Resource;
+import org.opennms.newts.api.Results.Row;
+import org.opennms.newts.api.SampleRepository;
+import org.opennms.newts.api.SampleSelectCallback;
+import org.opennms.newts.api.Timestamp;
+import org.opennms.newts.api.query.AggregationFunction;
+import org.opennms.newts.api.query.ResultDescriptor;
+import org.opennms.newts.api.query.StandardAggregationFunctions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.orm.ObjectRetrievalFailureException;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Used to retrieve measurements from {@link org.opennms.newts.api.SampleRepository}.
+ *
+ * If a request to {@link #fetch} spans multiple resources, separate calls to
+ * the {@link SampleRepository} will be performed in parallel.
+ *
+ * Reading the samples and computing the aggregated values can be very CPU intensive.
+ * The "parallelism" attribute is used to set an upper limit on how may concurrent threads
+ * can be used to perform these calculations. By default, this is set to the number of
+ * cores, but can be reduced if the operator wishes to ensure cores are available
+ * for other purposes.
+ *
+ * @author jwhite
+ */
+public class TimeseriesFetchStrategy implements MeasurementFetchStrategy {
```

Comment:
Can we refactor this to share the common parts with the Newts implementation? Seems like there is a lot of duplication here. I'd be OK with using the existing Newts systems properties for the constants.

Reply:
Today we have no dependency to the Newts module in Opennms (just references on the Newts library). I wonder if we shouldn't keep it that way?
It would give us the advantage of:
- refactor the fetch logic where needed
- over time lower and finally remove the dependency to the newts library (technically there is no need to use the newts value objects - we could just use the ones from the integration layer api.
I wonder of these advantages are not more that the code duplication for now? What do you think?
The same applies to the properties.